### PR TITLE
Add trace parsing from a file instead of realtime

### PR DESCRIPTION
**Pull request by acyr**
Add a new interface set_trace_filename() that will enable reading from an ETL file rather than real-time capture.
Also added a new NativeExample that demonstrates this capability with a sample ETL.

### CODE REVIEW

**Comment by acyr** [comment]
@microsoft-github-policy-service agree

File: examples/NativeExamples/examples.h [hunk 1]
```diff
@@ -8,6 +8,11 @@ struct kernel_and_user_trace_001
     static void start();
 };
 
+struct user_trace_009_from_file
```
Comment (swannman) [inline examples/NativeExamples/examples.h:11]: let's move this to the bottom of the file

File: examples/NativeExamples/NativeExamples.vcxproj [hunk 2]
```diff
@@ -152,6 +152,7 @@
     </Link>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClCompile Include="user_trace_009_from_file.cpp" />
```
Comment (swannman) [inline examples/NativeExamples/NativeExamples.vcxproj:155]: let's move this to the line after `user_trace_008_stacktrace.cpp` below

File: examples/NativeExamples/user_trace_009_from_file.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// This example shows how to use a user_trace and a kernel_trace in the same program.
```
Comment (swannman) [inline examples/NativeExamples/user_trace_009_from_file.cpp:4]: let's update this comment to reflect what this file is demonstrating

File: examples/NativeExamples/user_trace_009_from_file.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+#include <iostream>
+#include <thread>
+#include <condition_variable>
+#include "..\..\krabs\krabs.hpp"
+#include "examples.h"
+
+static void setup_ps_provider(krabs::provider<>& provider);
```
Comment (swannman) [inline examples/NativeExamples/user_trace_009_from_file.cpp:12]: are either of these being used?

File: examples/NativeExamples/user_trace_009_from_file.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        });
+
+    // the user_trace needs to know about the provider that we've set up.
+    trace.enable(provider);
+
+    // specify a filename to read from, will disable realtime and source from file instead
+    trace.set_trace_filename(L"..\\..\\examples\\NativeExamples\\powershell.etl");
+
+    // begin listening for events. This call blocks, so if you want to do other things
```
Comment (swannman) [inline examples/NativeExamples/user_trace_009_from_file.cpp:64]: what happens when the end of file is reached - does the trace session continue to block the thread?

File: krabs/krabs/etw.hpp [hunk 6]
```diff
@@ -262,9 +262,18 @@ namespace krabs { namespace details {
     EVENT_TRACE_LOGFILE trace_manager<T>::fill_logfile()
     {
         EVENT_TRACE_LOGFILE file = {};
-        file.LoggerName          = const_cast<wchar_t*>(trace_.name_.c_str());
-        file.ProcessTraceMode    = PROCESS_TRACE_MODE_EVENT_RECORD |
-                                   PROCESS_TRACE_MODE_REAL_TIME;
+
+        if (!trace_.logFilename_.empty())
```
Comment (swannman) [inline krabs/krabs/etw.hpp:266]: nit: let's maintain the original alignment on the `=` characters:

```
        if (!trace_.logFilename_.empty())
        {
            file.LogFileName         = const_cast<wchar_t*>(trace_.logFilename_.c_str());
            file.ProcessTraceMode    = PROCESS_TRACE_MODE_EVENT_RECORD;
        }
        else
        {
            file.LoggerName          = const_cast<wchar_t*>(trace_.name_.c_str());
            file.ProcessTraceMode    = PROCESS_TRACE_MODE_EVENT_RECORD |
                                       PROCESS_TRACE_MODE_REAL_TIME;
        }
```

**Review by swannman — CHANGES_REQUESTED** [review]
Thanks for submitting this! Added a few comments, mostly style-related.

File: examples/NativeExamples/user_trace_009_from_file.cpp (hunk 5 above)
Comment (acyr) [reply examples/NativeExamples/user_trace_009_from_file.cpp:64]: It unblocks and returns at EOF.  I can call stop, but I don't know if that's necessary... maybe better form?

File: krabs/krabs/etw.hpp (hunk 6 above)
Comment (acyr) [reply krabs/krabs/etw.hpp:266]: Will do.

File: examples/NativeExamples/user_trace_009_from_file.cpp (hunk 4 above)
Comment (acyr) [reply examples/NativeExamples/user_trace_009_from_file.cpp:12]: Whoops.  Nope, leftover garbage.  Originally user kernel_and_user_trace_001 as the base for the file.

File: examples/NativeExamples/user_trace_009_from_file.cpp (hunk 5 above)
Comment (acyr) [reply examples/NativeExamples/user_trace_009_from_file.cpp:64]: I'll add a stop() call for completeness, and update comment to reflect the EOF results in start() returning control.

**Review by swannman — APPROVED** [review]

**Comment by swannman** [comment]
@acyr do you want to bump the version number in the three `.nuspec` files in the root of the project so we can publish updated nugets?

**Review by swannman — APPROVED** [review]

**Pull request merged**